src/mat/impls/inv/matinv.c

```diff
@@ -98,29 +116,64 @@ static PetscErrorCode MatInvComputeNullSpace_Inv(Mat imat)
       TRY( MatCholeskyFactorNumeric(F,Kl,NULL) );
     }
     TRY( MatMumpsGetInfog(F,28,&defect) ); /* get numerical defect, i.e. number of null pivots encountered during factorization */
+    /* mumps->petsc_size > 1 implies mumps->id.ICNTL(21) = 1 (distributed solution ) */
+    mm = (defect && mumps->petsc_size > 1) ? mumps->id.lsol_loc : m;  /* = length of sol_loc = INFO(23) */
   }
 
-  TRY( MatCreateDensePermon(PETSC_COMM_SELF,m,defect,m,defect,NULL,&Rl) );
+  TRY( MatCreateDensePermon(blockComm,mm,PETSC_DECIDE,M,defect,NULL,&Rl) );
 
   if (defect) {
-    TRY( MatDenseGetArray(Rl,&array) );
-    mumps->id.rhs = (MumpsScalar*)array;
-    mumps->id.lrhs = m;
+    /* stash sol_loc allocated in MatFactorNumeric_MUMPS() */
+    MumpsScalar *sol_loc_orig = mumps->id.sol_loc;
+    MumpsScalar *array;
+
+    /* inject matrix array as sol_loc */
+    TRY( MatDenseGetArray(Rl,(MumpsScalar**)&array) );
+    if (mumps->petsc_size > 1) {
+      mumps->id.sol_loc = array;
+      if (!mumps->myid) {
+        /* Define dummy rhs on the host otherwise MUMPS fails with INFOG(1)=-22,INFOG(2)=7 */
+        TRY( PetscMalloc1(M,&mumps->id.rhs) );
+      }
+    } else mumps->id.rhs = array;
+    /* mumps->id.nrhs is reset by MatMatSolve_MUMPS()/MatSolve_MUMPS() */
     mumps->id.nrhs = defect;
     TRY( MatMumpsSetIcntl(F,25,-1) ); /* compute complete null space */
     mumps->id.job = JOB_SOLVE;
     PetscMUMPS_c(&mumps->id);
     if (mumps->id.INFOG(1) < 0) SETERRQ2(PETSC_COMM_SELF,PETSC_ERR_LIB,"Error reported by MUMPS in solve phase: INFOG(1)=%d,INFOG(2)=%d\n",mumps->id.INFOG(1),mumps->id.INFOG(2));
-    TRY( MatDenseRestoreArray(Rl,&array) );
+
+    if (mumps->petsc_size > 1 && !mumps->myid) {
+      TRY( PetscFree(mumps->id.rhs) );
+    }
     TRY( MatMumpsSetIcntl(F,25,0) ); /* perform a normal solution step next time */
+
+    /* restore matrix array */
+    TRY( MatDenseRestoreArray(Rl,(MumpsScalar**)&array) );
+    /* restore stashed sol_loc */
+    mumps->id.sol_loc = sol_loc_orig;
   }
 
-  TRY( MatCreateBlockDiag(PETSC_COMM_WORLD,Rl,&R) );
+  //TODO return just NULL if defect=0 ?
```

Comment:
I am for returning NULL when defect==0.

Reply:
I agree but let me finish this after or within PR #20 since we need to be careful with distinguishing between singular Hessian with R not yet set and non-singular Hessian (with empty R). That's why I left the TODO.

Reply:
https://github.com/permon/permon/pull/20#issuecomment-635186807